report/4.Solution/ParserImplementation.tex

```diff
@@ -115,7 +118,83 @@ \subsection{Implementing a recursive decent parser}
+            } else {
+                val lambda = parseLambdaBody(emptyList())
+                AstNode.Command.Expression.LambdaExpression(listOf(), lambda.type, lambda.lambdaBody)
+            }
+        is Token.Identifier -> {
+            val token = accept<Token.Identifier>()
+            val symbol = retrieveSymbol(token.value)
+            if (symbol.isFunctions) {
+        	    ...parse needed arguments...
+        	    AstNode.Command.Expression.FunctionCall(
+        	    AstNode.Command.Expression.Value.Identifier(token.value),
+        	    listOf(expression) + secondaryArguments
+            } else expression
+        } else -> expression
+\end{lstlisting}
+
+As seen in snippet \ref{lis:parsePotCall}, the method first determines if the current token is a lambda expression, signified by a \textit{blockStart-token} (3).
+Should it be a lambda expression, the parser determines whether or not there is a functions call within the lambda that needs to be parsed as well (7).
+If there is not a function-call within the lambda, the lambda expression itself is returned from the method (10). 
+Should the token be an identifier, the parser determines if the identifier signifies a function, in which case it is determined that a function-call is being made (16).
+The method looks for the needed arguments for the function, and parse them as expressions.
+The method then returns the function-call with its corresponding arguments (18).
+If the identifier does not signify a function, it is then parsed as a regular expression.
+
+All expressions that are determined to not be function-calls, are parsed by the method \textbf{\textit{parseExpressionAtomic}} seen in snippet \ref{lis:parseExpr}.
+
+\begin{lstlisting}[language=Kotlin,label=lis:parseExpr,caption=A simplified version of the parseExpressionAtomic method from the parser.]
+private fun parseExpressionAtomic(): AstExpression =
+    when (current.token) {
+        Token.SpecialChar.SquareBracketStart -> parseListDeclaration()
+        Token.SpecialChar.ParenthesesStart -> {
+            if (isLambdaParameters()) {
+                parseLambdaDeclaration()
+            } else {
+                if(upcomingTuple()){
+                    parseTupleExpression()
+                } else {
+                    parseExpression()
+                }
+            }
+        }
+        Token.Literal -> acceptLiteral()
+        Token.SpecialChar.Colon -> {
+        	accept<Token.SpecialChar.Colon>()
+        	val token = accept<Token.Identifier>()
+        	retrieveSymbol(token.value).handle(..handle token as function..)
+        }
+        Token.Identifier -> {
+            val token = accept<Token.Identifier>()
+            retrieveSymbol(token.value).handle(..Handle identifier..)
+        }
+        else -> unexpectedTokenError(current.token)
+    }
+            
+\end{lstlisting}
 
+The method itself deals with determining the type of the expression that is to be parsed, and then parsing it atomically.
```

Comment:
There is something going on with the tense on "parsing" in this sentence. Not sure what ;)

Reply:
Enig med Jonas. Ændr til
"The method determines ..... and then parses it ....."